File: Py/cs50/crack.c
#include <stdio.h>
#include <cs50.h>
#include <string.h>
#include <ctype.h>
#include <crypt.h>

int main(int argc, char *argv[]) 
{
    if (argc != 2)
    {
        printf("Usage: ./crack hash\n");
        return 1;
    }
    char newHash[6] = {};
    char guess[6] = {};
    char salt[3];
    salt[0] = argv[1][0];
    salt[1] = argv[1][1];
    char ascii[95];
    for (int i = 0; i < 95; i++)  
    {
        ascii[i] = (char)(i + 32);
    }
    for (int i = 0; i < 95; i++)
    {
        for (int j = 0; j < 95; j++)
        {
            for (int k = 0; k < 95; k++)
            {
                for (int l = 0; l < 95; l++)
                {
                    for (int m = 0; m < 95; m++)
                    {
                        newHash[0] = ascii[i];
                        newHash[1] = ascii[j];
                        newHash[2] = ascii[k];
                        newHash[3] = ascii[l];
                        newHash[4] = ascii[m];
                        for (int n = 0, o = 5; n <= 4; n++, o--)
                        {
                            for (int p = 0; p < n + 1; p++)
                            {
                                strncpy(guess, newHash + p, o);
                                guess[o] = '\0';
                                if (strcmp(crypt(guess, salt), argv[1]) == 0)
                                {
                                    printf("%s\n", guess);
                                    return 0;
                                }
                            }
                        }
                    }
                }
            }
        }
    }
}

    
        
            
                
                    
             

File: Py/cs50/pset6/crack.py
import sys
import crypt

# get hash from user
if len(sys.argv) != 2 or not sys.argv[1].isalnum():
    sys.exit("Usage: python crack hash")

# loop over all possible strings to length 4


def main():
    # generate asc
    asc = list(range(0, 1)) + list(range(65, 91)) + list(range(97, 123))

    # initialize password guess
    guess = ''
    for a in asc:
        for b in asc:
            for c in asc:
                for d in asc:
                    guess = guess if a == 0 else guess + chr(a)
                    guess = guess if b == 0 else guess + chr(b)
                    guess = guess if c == 0 else guess + chr(c)
                    guess = guess if d == 0 else guess + chr(d)
                    if compare(guess):
                        print(guess)
                        break
                    guess = ''
                    if compare(guess):
                        break
                if compare(guess):
                    break
            if compare(guess):
                break
        if compare(guess):
            break

# compare guess to hashedPass


def compare(guess):
    # specify salt
    hashedPass = sys.argv[1]
    salt = sys.argv[1][0:2]
    return crypt.crypt(guess, salt) == hashedPass


if __name__ == "__main__":
    main()